File: CPP/src/AgendaService.cpp
#include "AgendaService.hpp"
#include "Date.hpp"
#include <iostream>
#include <algorithm>
AgendaService::AgendaService() {
    startAgenda();
}
AgendaService::~AgendaService() {
    quitAgenda();
}
bool AgendaService::userLogIn(const std::string t_userName,
                                const std::string t_password) {
    auto func = [&t_userName, &t_password](const User& t_user)-> bool {
        return (t_user.getName() == t_userName &&
                    t_user.getPassword() == t_password);
    };
    return !m_storage->queryUser(func).empty();
}
bool AgendaService::userRegister(const std::string userName,
                                const std::string password,
                                const std::string email,
                                const std::string phone) {
    auto func = [&userName](const User& t_user) -> bool{
        return t_user.getName() == userName;
    };
    if (!m_storage->queryUser(func).empty()) { return false; }
    m_storage->createUser(User(userName, password, email, phone));
    return true;
}
bool AgendaService::deleteUser(const std::string userName,
                            const std::string password) {
    auto func = [&userName, &password](const User& user)->bool {
        return (user.getName() == userName && user.getPassword() == password);
    };
    if (m_storage->deleteUser(func) != 0) {
        auto func_delMeeting = [&userName](const Meeting& meeting)->bool {
            return (meeting.getSponsor() == userName ||
                    meeting.isParticipator(userName));
        };
        m_storage->deleteMeeting(func_delMeeting);
        return true;
    }
    return false;
}
std::list<User> AgendaService::listAllUsers() const {
    auto func = [](const User& per)->bool {
        return true;
    };
    return m_storage->queryUser(func);
}
bool AgendaService::createMeeting(const std::string userName,
                                    const std::string title,
                                    const std::string startDate,
                                    const std::string endDate,
                       const std::vector<std::string> participator) {
    // check if every participator is unique.
    if (participator.empty()) { return false; }
    if (!participator.empty()) {
        for (int i = 0; i < participator.size() - 1; i++) {
            for (int j = i + 1; j < participator.size(); j++) {
                    if (participator[i] == participator[j]) {
                        return false;
                    }
                }
        }
    }
    // check if the time is correct.
    if (!Date::isValid(Date::stringToDate(startDate))) {
        return false;
    }
    if (!Date::isValid(Date::stringToDate(endDate))) {
        return false;
    }
    if (Date::stringToDate(startDate)
            >= Date::stringToDate(endDate)) { 
        return false;
    }
    // check if the username has been register.
    auto checkSponsor = [&userName](const User& user)->bool {
        return user.getName() == userName;
    };
    if (m_storage->queryUser(checkSponsor).empty()) return false;
    // check if every participator has been registered.
    for (auto &part : participator) {
        auto checkParticipator = [&](const User& user) {
            return user.getName() == part;
        };
        if (m_storage->queryUser(checkParticipator).empty())
            return false;
    }

    auto func = [&userName, &startDate, &endDate, &title](const Meeting& t_meeting) -> bool {
        // check if the title has been used.
        if (title == t_meeting.getTitle()) { return true; }
        // check if the sponsor is vacant.
        if (t_meeting.getSponsor() == userName || t_meeting.isParticipator(userName)) {
            return (!(Date::stringToDate(startDate) >= t_meeting.getEndDate()
                    || Date::stringToDate(endDate) <= t_meeting.getStartDate()));
        }
        return false;
    };
    if (!m_storage->queryMeeting(func).empty()) { return false; }
    // check if every participator is vacant.
    for (std::string t_part : participator) {
        if (t_part == userName) { return false; }
        auto func = [&t_part, &startDate, &endDate](const Meeting& t_meeting) -> bool {
            if (t_meeting.getSponsor() == t_part || t_meeting.isParticipator(t_part)) {
                return (!(Date::stringToDate(startDate) >= t_meeting.getEndDate() ||
                        Date::stringToDate(endDate) <= t_meeting.getStartDate()));
            }
            return false;
        };
        if (!m_storage->queryMeeting(func).empty()) { return false; }
    }
    m_storage->createMeeting(
            Meeting(userName, participator, startDate, endDate, title));
    return true;
}
std::list<Meeting> AgendaService::meetingQuery(const std::string userName,
                                    const std::string title) const {
    auto func = [&userName, &title](const Meeting& t_meeting)->bool {
        if (t_meeting.getTitle() == title) {
            return (t_meeting.getSponsor() == userName ||
                    t_meeting.isParticipator(userName));
        }
        return false;
    };
    return m_storage->queryMeeting(func);
}
std::list<Meeting> AgendaService::meetingQuery(const std::string userName,
                                const std::string startDate,
                                const std::string endDate) const {
    // check if the time is valid.
    if (!(Date::isValid(Date::stringToDate(startDate)) &&
            Date::isValid(Date::stringToDate(endDate)))) {
        return std::list<Meeting>();
    }
    if (Date::stringToDate(startDate) >
            Date::stringToDate(endDate)) { return std::list<Meeting>(); }

    auto func = [&userName, &startDate, &endDate](const Meeting & t_meeting)->bool {
        if (t_meeting.getSponsor() == userName ||
                t_meeting.isParticipator(userName)) {
            return (!(Date::stringToDate(endDate) < t_meeting.getStartDate() || 
                    Date::stringToDate(startDate) > t_meeting.getEndDate()));
        }
        return false;
    };
    return m_storage->queryMeeting(func);
}
std::list<Meeting> AgendaService::listAllMeetings(
                    const std::string userName) const {
    auto func = [&userName](const Meeting& t_meeting)->bool {
        return (t_meeting.getSponsor() == userName ||
                t_meeting.isParticipator(userName));
    };
    return m_storage->queryMeeting(func);
}
std::list<Meeting> AgendaService::listAllSponsorMeetings(
                const std::string userName) const {
    auto func = [&userName](const Meeting& t_meeting)->bool {
        return (t_meeting.getSponsor() == userName);
    };
    return m_storage->queryMeeting(func);
}
std::list<Meeting> AgendaService::listAllParticipateMeetings(
                const std::string userName) const {
    auto func = [&userName](const Meeting& t_meeting)->bool {
        return t_meeting.isParticipator(userName);
    };
    return m_storage->queryMeeting(func);
}
bool AgendaService::deleteMeeting(const std::string userName,
                                    const std::string title) {
    auto func = [&userName, &title](const Meeting& t_meeting)->bool {
        return (t_meeting.getSponsor() == userName &&
                t_meeting.getTitle() == title);
    };
    return m_storage->deleteMeeting(func);
}
bool AgendaService::deleteAllMeetings(const std::string userName) {
    auto func = [&userName](const Meeting& t_meeting)->bool {
        return t_meeting.getSponsor() == userName;
    };
    return m_storage->deleteMeeting(func) != 0;
}
void AgendaService::startAgenda() {
    m_storage = Storage::getInstance();
}
void AgendaService::quitAgenda() {
    m_storage->sync();
}

File: Agenda/src/src/AgendaService.java
/**
 * Created by yanzexin on 16/8/30.
 * All right reserved.
 */
package src;
import java.util.*;

public class AgendaService {
    private Storage m_storage;
    public AgendaService() {
        m_storage = Storage.getInstance();
    }
    public boolean userLogIn(final String t_userName, final String t_password) {
        return m_storage.logIn(t_userName, t_password);
    }
    public boolean userRegister(final String t_userName, final String t_password, final String t_email, final String t_phone) {
        if (m_storage.existUser(t_userName)) {
            return false;
        }
        m_storage.createUser(new User(t_userName, t_password, t_email, t_phone));
        return true;
    }
    public boolean deleteUser(final String t_userName, final String t_password) {
        m_storage.deleteUser(t_userName, t_password);
        return !m_storage.existUser(t_userName);
	}
    public boolean changePassword(final String t_userName, final String old_password, final String new_password) {
        return m_storage.changePassword(t_userName, old_password, new_password);
    }
    public boolean changeEmail(final String t_userName, final String t_password, final String t_email) {
        return m_storage.changeEmail(t_userName, t_password, t_email);
    }
    public boolean changePhone(final String t_userName, final String t_password, final String t_phone) {
        return m_storage.changePhone(t_userName, t_password, t_phone);
    }
    public boolean createMeeting(final String t_Sponsor, final String t_title, final String t_StartDate, final String t_EndDate, final ArrayList<String> t_participator) {
        /**
         * check if exist participator or
         * if every participator is unique.
         */
        if (t_participator.size() == 0) { return false; }
        else if (t_participator.size() != new HashSet<>(t_participator).size()) {
            return false;
        }
        if (!Date.isValid(new Date(t_StartDate))) {
            return false;
        }
        if (!Date.isValid(new Date(t_EndDate))) {
            return false;
        }
        if (Date.stringToDate(t_StartDate).compare(Date.stringToDate(t_EndDate)) ||
            Date.stringToDate(t_StartDate).equals(Date.stringToDate(t_EndDate))) {
            return false;
        }
        if (!m_storage.existUser(t_Sponsor)) {
            return false;
        }
        for (String perParticipator : t_participator) {
            if (!m_storage.existUser(perParticipator)) {
                return false;
            }
        }
        if (m_storage.queryMeeting(t_meeting -> {
            if (t_meeting.getTitle().equals(t_title)) { return true; }
            if (t_meeting.getSponsor().equals(t_Sponsor) || t_meeting.isParticipator(t_Sponsor)) {
                return (!(t_StartDate.compareTo(Date.dateToString(t_meeting.getEndDate())) >= 0 ||
                        Date.dateToString(t_meeting.getStartDate()).compareTo(t_EndDate) >= 0));
            }
            return false;
        }).size() != 0) {
            return false;
        }
        for (String perParticipator : t_participator) {
            if (perParticipator.equals(t_Sponsor)) {
                return false;
            }
            if (m_storage.queryMeeting(t_meeting -> {
                if (t_meeting.getSponsor().equals(perParticipator) || t_meeting.isParticipator(perParticipator)) {
                    return (!(t_StartDate.compareTo(Date.dateToString(t_meeting.getEndDate())) >= 0 ||
                            Date.dateToString(t_meeting.getStartDate()).compareTo(t_EndDate) >= 0));
                }
                return false;
            }).size() != 0) {
                return false;
            }
        }
        m_storage.createMeeting(t_Sponsor,t_title,t_StartDate, t_EndDate, t_participator);
        return true;
    }
	 public final ArrayList<User> listAllUsers() {
        return m_storage.listAllUsers();
	}
    public final ArrayList<Meeting> listAllMeetings(final String t_userName) {
        return m_storage.queryMeeting(new MeetingFilter() {
            @Override
            public boolean filter(Meeting t_meeting) {
                return (t_meeting.getSponsor().equals(t_userName) || t_meeting.isParticipator(t_userName));
            }
        });
    }
    public final ArrayList<Meeting> listSponsorMeetings(final String t_userName) {
        return m_storage.queryMeeting(new MeetingFilter() {
            @Override
            public boolean filter(Meeting t_meeting) {
                return t_meeting.getSponsor().equals(t_userName);
            }
        });
    }
    public final ArrayList<Meeting> listAllParticipatorMeetings(final String t_userName) {
        return m_storage.queryMeeting(new MeetingFilter() {
            @Override
            public boolean filter(Meeting t_meeting) {
                return t_meeting.isParticipator(t_userName);
            }
        });
    }
    public final ArrayList<Meeting> queryMeetingByTitle(final String t_userName, final String t_title){
        return m_storage.queryMeeting(new MeetingFilter() {
            @Override
            public boolean filter(Meeting t_meeting) {
                if (t_meeting.getSponsor().equals(t_userName) || t_meeting.isParticipator(t_userName)) {
                    return t_meeting.getTitle().equals(t_title);
                }
                return false;
            }
        });
    }
    public final ArrayList<Meeting> queryMeetingByInterval(final String t_userName, final String t_startDate, final String t_endDate) {
        if (!Date.isValid(Date.stringToDate(t_startDate))) {
            return null;
        }
        if (!Date.isValid(Date.stringToDate(t_endDate))) {
            return null;
        }
        if (t_startDate.compareTo(t_endDate) > 0) {
            return null;
        }
        return m_storage.queryMeeting(new MeetingFilter() {
            @Override
            public boolean filter(Meeting t_meeting) {
                if (t_meeting.getSponsor().equals(t_userName) || t_meeting.isParticipator(t_userName)) {
                    return !(t_endDate.compareTo(Date.dateToString(t_meeting.getEndDate())) < 0 ||
                            t_startDate.compareTo(Date.dateToString(t_meeting.getStartDate())) > 0);
                }
                return false;
            }
        });
    }
    public void startAgenda() {
        m_storage = Storage.getInstance();
    }
    public static void main(String[] args) {
        AgendaService m_storage = new AgendaService();
        System.out.print(m_storage.userLogIn("stary", "123"));
    }
}
